### PR TITLE
Slight speed improvement in $get_hash() for RDS storrs

### PR DESCRIPTION
**Pull request by wlandau**
- Pre-compute the hash length and save it in the `storr` alongside `hash_algorithm`.
- Use `readChar()` instead of `readLines()` in `get_hash()`.
- Benchmarks: https://github.com/richfitz/storr/issues/96#issuecomment-452357226

### CODE REVIEW

**Comment by wlandau** [comment]
The Appveyor build failed because `rlang` just got updated on CRAN. I will refresh this in a couple days.

**Comment by richfitz** [comment]
Try updating the `appveyor.yml` to include

```yaml
environment:
  global:
    USE_RTOOLS: true
```

and appveyor will be able to install rlang

File: R/driver_rds.R [hunk 1]
```diff
@@ -186,6 +187,10 @@ R6_driver_rds <- R6::R6Class(
       }
       self$hash_algorithm <- driver_rds_config(path, "hash_algorithm",
                                                hash_algorithm, "md5", TRUE)
+
+      self$hash_length <- nchar(
```
Comment (richfitz) [inline R/driver_rds.R]: This looks like the sort of thing that will cause me a lot of confusion in future (all of the hash stuff needs an overhaul at some point).  An alternative approach would be an integer vector in `hash.R` that contains the lengths - they won't change!

```r
hash_length = c(md5 = 32L, ...)
```

and just look this up here (`self$hash_length <- hash_length[[self$hash_algorithm]]`)

Or just run `nchar(digest::digest(as.raw(0x00), self$hash_algorithm, serialize = FALSE))`

File: R/driver_rds.R [hunk 2]
```diff
@@ -197,7 +202,11 @@ R6_driver_rds <- R6::R6Class(
     },
 
     get_hash = function(key, namespace) {
-      readLines(self$name_key(key, namespace))
+      readChar(
```
Comment (richfitz) [inline R/driver_rds.R]: Format on one line please, without argument names

**Comment by richfitz** [comment]
Would you mind also extending the benchmark to show what this looks like end-to-end; it's not a lot of complexity but it would be nice to see the improvement (and the potential C improvement) in the context of the total time that storr takes to interact with these files

File: R/driver_rds.R (hunk 1 above)
Comment (wlandau) [reply R/driver_rds.R]: Sure, I like `nchar(digest::digest(...))`. See 21b2966c1e7d8aa055804bb3a6eea382f5c3d4f5.

File: R/driver_rds.R (hunk 2 above)
Comment (wlandau) [reply R/driver_rds.R]: Sure: bb2c63b2e8c2089421fc2dc47da162e853b47a6d.

**Comment by wlandau** [comment]
Re https://github.com/richfitz/storr/pull/97#issuecomment-452378223, I will extend the benchmark to compare this PR to existing `storr`. I will benchmark the C++-powered solution when I have a PR for that one.

**Comment by wlandau** [comment]
Using 27508231b3c061afed9bb243d4422f81291f2e94:

``` r
s <- storr::storr_rds(tempfile())
s$set("a", 1)  
microbenchmark::microbenchmark(
  get = s$get("a"),
  get_hash = s$get_hash("a")
)
#> Unit: microseconds
#>      expr     min       lq      mean  median       uq     max neval cld
#>       get 103.214 105.6915 111.24496 107.936 112.4715 144.784   100   b
#>  get_hash  78.081  81.7500  85.44661  83.557  86.0225 174.059   100  a
```

Using bb2c63b2e8c2089421fc2dc47da162e853b47a6d:

``` r
s <- storr::storr_rds(tempfile())
s$set("a", 1)  
microbenchmark::microbenchmark(
  get = s$get("a"),
  get_hash = s$get_hash("a")
)
#> Unit: microseconds
#>      expr    min      lq      mean  median      uq     max neval cld
#>       get 93.409 96.7315 101.68020 99.3240 103.639 154.736   100   b
#>  get_hash 70.102 73.3935  79.41743 75.0595  78.381 258.517   100  a
```

Hmm... that's a bit disappointing. If you like, we can hold off for a bit. I am planning to write a C++ alternative anyway.

**Comment by richfitz** [comment]
Closed in favour of #98 - larger changes but better performance gain

**Pull request closed without merging**